File: ios/ClimaAI/Services/APIClient.swift
import Foundation

class APIClient {
    static let shared = APIClient()
    
    private let baseURL: String
    private var accessToken: String?
    
    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()
    
    private let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.keyEncodingStrategy = .convertToSnakeCase
        return encoder
    }()
    
    private init() {
        #if DEBUG
        self.baseURL = "http://localhost:8000"
        #else
        self.baseURL = "https://api.climaai.com" // Replace with production URL
        #endif
    }
    
    func setAccessToken(_ token: String?) {
        self.accessToken = token
        if let token = token {
            UserDefaults.standard.set(token, forKey: "access_token")
        } else {
            UserDefaults.standard.removeObject(forKey: "access_token")
        }
    }
    
    func loadAccessToken() {
        self.accessToken = UserDefaults.standard.string(forKey: "access_token")
    }
    
    private func createRequest(
        endpoint: String,
        method: String = "GET",
        body: Data? = nil,
        requiresAuth: Bool = false
    ) throws -> URLRequest {
        guard let url = URL(string: baseURL + endpoint) else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if requiresAuth, let token = accessToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        if let body = body {
            request.httpBody = body
        }
        
        return request
    }
    
    private func performRequest<T: Decodable>(
        _ request: URLRequest
    ) async throws -> T {
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        if httpResponse.statusCode == 401 {
            // Token expired, clear it
            setAccessToken(nil)
            throw NSError(domain: "APIClient", code: 401, userInfo: [NSLocalizedDescriptionKey: "Authentication required"])
        }
        
        guard 200...299 ~= httpResponse.statusCode else {
            // Try to decode error
            if let error = try? decoder.decode(APIError.self, from: data) {
                throw error
            }
            throw NSError(
                domain: "APIClient",
                code: httpResponse.statusCode,
                userInfo: [NSLocalizedDescriptionKey: "Server error: \(httpResponse.statusCode)"]
            )
        }
        
        return try decoder.decode(T.self, from: data)
    }
    
    // MARK: - Generic REST surface
    //
    // AuthViewModel and AIInsightsViewModel were written against a generic
    // client (get/post/put/delete plus token helpers) that this class never
    // exposed — every call site failed to compile with "value of type
    // 'APIClient' has no member". These are thin wrappers over the same
    // createRequest/performRequest plumbing the specific endpoints below use, so
    // there is one code path for auth, error handling and 401 token clearing.

    /// True when an access token is held, without exposing the token itself.
    var isAuthenticated: Bool {
        accessToken != nil
    }

    /// Persist a token. Alias for `setAccessToken` in the vocabulary the view
    /// models already use.
    func saveToken(_ token: String) {
        setAccessToken(token)
    }

    /// Drop the stored token, e.g. on sign-out.
    func clearToken() {
        setAccessToken(nil)
    }

    func get<T: Decodable>(_ endpoint: String, requiresAuth: Bool = true) async throws -> T {
        let request = try createRequest(endpoint: endpoint, method: "GET", requiresAuth: requiresAuth)
        return try await performRequest(request)
    }

    func post<T: Decodable>(
        _ endpoint: String,
        body: Encodable? = nil,
        requiresAuth: Bool = true
    ) async throws -> T {
        let data = try body.map { try JSONEncoder().encode(AnyEncodable($0)) }
        let request = try createRequest(
            endpoint: endpoint, method: "POST", body: data, requiresAuth: requiresAuth
        )
        return try await performRequest(request)
    }

    func put<T: Decodable>(
        _ endpoint: String,
        body: Encodable? = nil,
        requiresAuth: Bool = true
    ) async throws -> T {
        let data = try body.map { try JSONEncoder().encode(AnyEncodable($0)) }
        let request = try createRequest(
            endpoint: endpoint, method: "PUT", body: data, requiresAuth: requiresAuth
        )
        return try await performRequest(request)
    }

    func delete<T: Decodable>(_ endpoint: String, requiresAuth: Bool = true) async throws -> T {
        let request = try createRequest(endpoint: endpoint, method: "DELETE", requiresAuth: requiresAuth)
        return try await performRequest(request)
    }

    // MARK: - Auth Endpoints
    
    func register(email: String, password: String, fullName: String?, deviceToken: String?) async throws -> TokenResponse {
        let registerData = UserRegister(
            email: email,
            password: password,
            fullName: fullName,
            platform: "ios",
            deviceToken: deviceToken
        )
        
        let body = try encoder.encode(registerData)
        let request = try createRequest(endpoint: "/users/register", method: "POST", body: body)
        
        let response: TokenResponse = try await performRequest(request)
        setAccessToken(response.accessToken)
        return response
    }
    
    func login(email: String, password: String) async throws -> TokenResponse {
        let loginData = UserLogin(email: email, password: password)
        let body = try encoder.encode(loginData)
        let request = try createRequest(endpoint: "/users/login", method: "POST", body: body)
        
        let response: TokenResponse = try await performRequest(request)
        setAccessToken(response.accessToken)
        return response
    }
    
    func getCurrentUser() async throws -> User {
        let request = try createRequest(endpoint: "/users/me", requiresAuth: true)
        return try await performRequest(request)
    }
    
    // MARK: - Weather Endpoints
    
    func getWeather(latitude: Double, longitude: Double) async throws -> WeatherResponse {
        let endpoint = "/weather/current?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: false)
        return try await performRequest(request)
    }
    
    func getHourlyForecast(latitude: Double, longitude: Double, hours: Int = 24) async throws -> WeatherResponse {
        let endpoint = "/weather/hourly?latitude=\(latitude)&longitude=\(longitude)&hours=\(hours)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: false)
        return try await performRequest(request)
    }
    
    func getDailyForecast(latitude: Double, longitude: Double, days: Int = 7) async throws -> WeatherResponse {
        let endpoint = "/weather/daily?latitude=\(latitude)&longitude=\(longitude)&days=\(days)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: false)
        return try await performRequest(request)
    }
    
    func getAirQuality(latitude: Double, longitude: Double) async throws -> AirQuality {
        let endpoint = "/weather/air-quality?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: false)
        
        struct AQResponse: Codable {
            let airQuality: AirQuality
            
            enum CodingKeys: String, CodingKey {
                case airQuality = "air_quality"
            }
        }
        
        let response: AQResponse = try await performRequest(request)
        return response.airQuality
    }
    
    // MARK: - AI Endpoints (Premium)
    
    func getAIInsights(latitude: Double, longitude: Double, locationName: String = "your location") async throws -> AIInsightsResponse {
        let endpoint = "/ai/insights?latitude=\(latitude)&longitude=\(longitude)&location_name=\(locationName.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? locationName)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func getDailySummary(latitude: Double, longitude: Double, locationName: String = "your location") async throws -> DailySummary {
        let endpoint = "/ai/summary?latitude=\(latitude)&longitude=\(longitude)&location_name=\(locationName.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? locationName)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        
        struct Response: Codable {
            let summary: DailySummary
        }
        
        let response: Response = try await performRequest(request)
        return response.summary
    }
    
    func getOutfitRecommendation(latitude: Double, longitude: Double) async throws -> OutfitRecommendation {
        let endpoint = "/ai/outfit?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func getActivityRecommendations(latitude: Double, longitude: Double) async throws -> [ActivityRecommendation] {
        let endpoint = "/ai/activities?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func getHealthInsights(latitude: Double, longitude: Double) async throws -> HealthInsight {
        let endpoint = "/ai/health?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        return try await performRequest(request)
    }
    
    // MARK: - Subscription Endpoints
    
    func getSubscriptionStatus() async throws -> SubscriptionStatus {
        let request = try createRequest(endpoint: "/subscriptions/status", requiresAuth: true)
        return try await performRequest(request)
    }
    
    func startTrial(receiptData: String) async throws -> Subscription {
        struct TrialRequest: Codable {
            let platform: String
            let plan: String
            let receiptData: String
            
            enum CodingKeys: String, CodingKey {
                case platform, plan
                case receiptData = "receipt_data"
            }
        }
        
        let trialData = TrialRequest(platform: "apple", plan: "monthly", receiptData: receiptData)
        let body = try encoder.encode(trialData)
        let request = try createRequest(endpoint: "/subscriptions/trial", method: "POST", body: body, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func activateSubscription(plan: String, receiptData: String) async throws -> Subscription {
        struct ActivateRequest: Codable {
            let platform: String
            let plan: String
            let receiptData: String
            
            enum CodingKeys: String, CodingKey {
                case platform, plan
                case receiptData = "receipt_data"
            }
        }
        
        let activateData = ActivateRequest(platform: "apple", plan: plan, receiptData: receiptData)
        let body = try encoder.encode(activateData)
        let request = try createRequest(endpoint: "/subscriptions/activate", method: "POST", body: body, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func getSubscriptionPlans() async throws -> [SubscriptionPlan] {
        let request = try createRequest(endpoint: "/subscriptions/plans")
        
        struct PlansResponse: Codable {
            let plans: [SubscriptionPlan]
        }
        
        let response: PlansResponse = try await performRequest(request)
        return response.plans
    }
    
    // MARK: - Alerts Endpoints
    
    func getAlerts(latitude: Double, longitude: Double, locationName: String = "your location") async throws -> AlertsResponse {
        let endpoint = "/alerts?latitude=\(latitude)&longitude=\(longitude)&location_name=\(locationName.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? locationName)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func getAlertHistory() async throws -> AlertHistoryResponse {
        let request = try createRequest(endpoint: "/alerts/history", requiresAuth: true)
        return try await performRequest(request)
    }
    
    func dismissAlert(alertId: Int) async throws -> MessageResponse {
        let request = try createRequest(endpoint: "/alerts/\(alertId)/dismiss", method: "POST", requiresAuth: true)
        return try await performRequest(request)
    }
    
    // MARK: - Health Endpoints (Pollen, Flu Risk, Migraine, Activities)
    
    func getPollenForecast(latitude: Double, longitude: Double, days: Int = 5) async throws -> PollenForecastResponse {
        let endpoint = "/health/pollen?latitude=\(latitude)&longitude=\(longitude)&days=\(days)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getTodaysPollen(latitude: Double, longitude: Double) async throws -> PollenTodayResponse {
        let endpoint = "/health/pollen/today?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getFluRisk(latitude: Double, longitude: Double) async throws -> FluRiskResponse {
        let endpoint = "/health/flu-risk?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getMigraineRisk(latitude: Double, longitude: Double) async throws -> MigraineRiskResponse {
        let endpoint = "/health/migraine-risk?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getAllActivities(latitude: Double, longitude: Double) async throws -> AllActivitiesResponse {
        let endpoint = "/health/activities?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getActivityForecast(activity: String, latitude: Double, longitude: Double) async throws -> ActivityDetailResponse {
        let endpoint = "/health/activities/\(activity)?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    // MARK: - Notification Endpoints
    
    func registerDevice(token: String, platform: String = "ios", deviceInfo: [String: Any]? = nil) async throws -> DeviceRegisterResponse {
        var body: [String: Any] = ["token": token, "platform": platform]
        if let info = deviceInfo { body["device_info"] = info }
        let bodyData = try JSONSerialization.data(withJSONObject: body)
        let request = try createRequest(endpoint: "/notifications/register-device", method: "POST", body: bodyData, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func unregisterDevice(token: String) async throws -> MessageResponse {
        let body = ["token": token]
        let bodyData = try encoder.encode(body)
        let request = try createRequest(endpoint: "/notifications/unregister-device", method: "POST", body: bodyData, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func getNotificationPreferences() async throws -> NotificationPreferencesWrapper {
        let request = try createRequest(endpoint: "/notifications/preferences", requiresAuth: true)
        return try await performRequest(request)
    }
    
    func updateNotificationPreferences(weatherAlerts: Bool? = nil, dailySummary: Bool? = nil, severeWeather: Bool? = nil) async throws -> NotificationPreferencesUpdateResponse {
        var body: [String: Bool] = [:]
        if let wa = weatherAlerts { body["weather_alerts"] = wa }
        if let ds = dailySummary { body["daily_summary"] = ds }
        if let sw = severeWeather { body["severe_weather"] = sw }
        let bodyData = try JSONSerialization.data(withJSONObject: body)
        let request = try createRequest(endpoint: "/notifications/preferences", method: "PUT", body: bodyData, requiresAuth: true)
        return try await performRequest(request)
    }
    
    func sendTestNotification(title: String = "Test", message: String = "Test notification") async throws -> TestNotificationResponse {
        let body = ["title": title, "message": message]
        let bodyData = try encoder.encode(body)
        let request = try createRequest(endpoint: "/notifications/test", method: "POST", body: bodyData, requiresAuth: true)
        return try await performRequest(request)
    }
    
    // MARK: - Personalization Endpoints
    
    func trackEvent(eventType: String, eventData: [String: Any], weatherContext: [String: Any]? = nil) async throws -> MessageResponse {
        var body: [String: Any] = ["event_type": eventType, "event_data": eventData]
        if let ctx = weatherContext { body["weather_context"] = ctx }
        let bodyData = try JSONSerialization.data(withJSONObject: body)
        let request = try createRequest(endpoint: "/personalization/track", method: "POST", body: bodyData)
        return try await performRequest(request)
    }
    
    func getPersonalizationProfile(userId: String = "demo_user") async throws -> UserPreferenceProfileResponse {
        let endpoint = "/personalization/profile?user_id=\(userId)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getPersonalizedRecommendations(temperature: Double, humidity: Int, uvIndex: Double, precipProbability: Int) async throws -> PersonalizedContentResponse {
        let endpoint = "/personalization/recommendations?temperature=\(temperature)&humidity=\(humidity)&uv_index=\(uvIndex)&precipitation_probability=\(precipProbability)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func shouldNotify(notificationType: String, userId: String = "demo_user") async throws -> ShouldNotifyResponse {
        let endpoint = "/personalization/should-notify?notification_type=\(notificationType)&user_id=\(userId)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    // MARK: - Precipitation Nowcast
    
    func getPrecipitationNowcast(latitude: Double, longitude: Double) async throws -> PrecipitationNowcastResponse {
        let endpoint = "/api/v1/weather/nowcast?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    // MARK: - NWS Weather Alerts
    
    func getWeatherAlertsNWS(latitude: Double, longitude: Double) async throws -> NWSAlertsResponse {
        let endpoint = "/api/weather/alerts?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getStateWeatherAlerts(state: String) async throws -> NWSStateAlertsResponse {
        let endpoint = "/api/weather/alerts/\(state)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    // MARK: - Multi-Source Weather
    
    func getMultiSourceWeather(latitude: Double, longitude: Double, sources: String? = nil) async throws -> MultiSourceWeatherResponse {
        var endpoint = "/api/weather/multi-source?latitude=\(latitude)&longitude=\(longitude)"
        if let s = sources { endpoint += "&sources=\(s)" }
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getUVIndex(latitude: Double, longitude: Double) async throws -> UVIndexData {
        let endpoint = "/api/weather/uv?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getMarineWeather(latitude: Double, longitude: Double) async throws -> MarineWeatherData {
        let endpoint = "/api/weather/marine?latitude=\(latitude)&longitude=\(longitude)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    func getHistoricalWeather(latitude: Double, longitude: Double, startDate: String, endDate: String) async throws -> HistoricalWeatherData {
        let endpoint = "/api/weather/historical?latitude=\(latitude)&longitude=\(longitude)&start_date=\(startDate)&end_date=\(endDate)"
        let request = try createRequest(endpoint: endpoint)
        return try await performRequest(request)
    }
    
    // MARK: - Travel Risk (Premium)
    
    func getTravelRisk(latitude: Double, longitude: Double, destination: String = "your destination") async throws -> TravelRiskAnalysis {
        let endpoint = "/api/travel-risk?latitude=\(latitude)&longitude=\(longitude)&destination=\(destination.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? destination)"
        let request = try createRequest(endpoint: endpoint, requiresAuth: true)
        return try await performRequest(request)
    }
}

/// Type-erasing box so the generic verbs above can take `Encodable` bodies.
/// `Encodable` alone cannot be passed to `JSONEncoder.encode`, which requires a
/// concrete conforming type.
private struct AnyEncodable: Encodable {
    private let encodeTo: (Encoder) throws -> Void

    init(_ wrapped: Encodable) {
        self.encodeTo = wrapped.encode
    }

    func encode(to encoder: Encoder) throws {
        try encodeTo(encoder)
    }
}

File: android/app/src/main/kotlin/com/climaai/app/data/ApiClient.kt
package com.climaai.app.data

import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.concurrent.TimeUnit
import com.climaai.app.BuildConfig
import com.climaai.app.data.api.*

interface ClimaAIApi {
    
    // Weather endpoints
    // The backend has no /api/weather route and names its parameters
    // latitude/longitude, not lat/lon — this returned 404 on every call.
    // getCurrentWeather below declares the same endpoint correctly but nothing
    // used it; WeatherRepository calls this one.
    @GET("/api/weather/current")
    suspend fun getWeather(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<WeatherResponse>
    
    @GET("/api/weather/current")
    suspend fun getCurrentWeather(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<WeatherResponse>
    
    @GET("/api/weather/hourly")
    suspend fun getHourlyForecast(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("hours") hours: Int = 24
    ): Response<WeatherResponse>
    
    @GET("/api/weather/daily")
    suspend fun getDailyForecast(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("days") days: Int = 7
    ): Response<WeatherResponse>
    
    @GET("/api/weather/air-quality")
    suspend fun getAirQuality(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<Map<String, AirQuality>>
    
    @GET("/api/weather/nowcast")
    suspend fun getNowcast(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<Map<String, Any>>
    
    // The router mounts the frame list at /radar/frames. /radar on its own is
    // only the shared prefix of that and the tile route, so it 404s.
    @GET("/api/weather/radar/frames")
    suspend fun getRadarFrames(): Response<Map<String, Any>>
    
    @GET("/api/weather/alerts")
    suspend fun getWeatherAlertsNWS(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<NWSAlertsResponse>
    
    @GET("/api/weather/alerts/{state}")
    suspend fun getStateWeatherAlerts(
        @Path("state") state: String
    ): Response<NWSStateAlertsResponse>
    
    // AI endpoints (Premium)
    // The router names these latitude/longitude like every other endpoint here.
    // Sending lat/lon made FastAPI reject the call with a 422 for a missing
    // required query parameter — the error the AI Insights card led to, and one
    // no amount of retrying could clear.
    @GET("/api/insights")
    suspend fun getAIInsights(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<AIInsightsResponse>
    
    @GET("/api/summary")
    suspend fun getDailySummary(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("location_name") locationName: String = "your location"
    ): Response<Map<String, Any>>
    
    @GET("/api/outfit")
    suspend fun getOutfitRecommendation(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<OutfitRecommendation>
    
    @GET("/api/activities")
    suspend fun getActivityRecommendations(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<List<ActivityRecommendation>>
    
    @GET("/api/health")
    suspend fun getHealthInsights(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<HealthInsight>
    
    @GET("/api/travel-risk")
    suspend fun getTravelRisk(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("destination") destination: String = "your destination"
    ): Response<TravelRiskAnalysis>
    
    // Auth endpoints
    @POST("/api/auth/register")
    suspend fun register(@Body user: UserRegister): Response<TokenResponse>
    
    // The router reads a JSON UserLogin body. This was declared as an OAuth2
    // style form post with a "username" field, so FastAPI rejected it with a 422
    // before it ever looked at the credentials — every sign-in failed, and the
    // screen reported it as bad credentials.
    @POST("/api/auth/login")
    suspend fun loginRequest(@Body credentials: UserLogin): Response<TokenResponse>

    @POST("/api/auth/forgot-password")
    suspend fun forgotPassword(@Body request: ForgotPasswordRequest): Response<Map<String, String>>
    
    @GET("/api/auth/me")
    suspend fun getCurrentUser(): Response<User>
    
    // There is no /api/users router at all; preferences are saved through the
    // auth router's user update, which expects them nested under "preferences".
    @PUT("/api/auth/me")
    suspend fun updatePreferences(@Body update: UserUpdateRequest): Response<User>

    // Account deletion (Play requires an in-app path now that registration is
    // in-app). Same /me as the profile — the users router mounts at /api/auth.
    // Replies 204 No Content.
    @DELETE("/api/auth/me")
    suspend fun deleteAccount(): Response<Unit>

    // Subscription endpoints
    @GET("/api/subscriptions/status")
    suspend fun getSubscriptionStatus(): Response<SubscriptionStatus>
    
    @POST("/api/subscriptions/validate")
    suspend fun validateReceipt(
        @Body data: ReceiptValidationRequest
    ): Response<ReceiptValidationResponse>
    
    @POST("/api/subscriptions/trial")
    suspend fun startTrial(@Body data: TrialRequest): Response<Subscription>

    // Server-side entitlement for a completed Play purchase: the backend
    // validates the token with Google and opens the subscription row that
    // /status reads. Without this call a purchase exists only on the device.
    @POST("/api/subscriptions/activate")
    suspend fun activateSubscription(
        @Body data: SubscriptionActivateRequest
    ): Response<Subscription>

    @GET("/api/subscriptions/plans")
    suspend fun getPlans(): Response<PlansResponse>
    
    // ============================================================
    // Alerts endpoints
    // ============================================================
    
    @GET("/alerts")
    suspend fun getAlerts(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("location_name") locationName: String = "your location"
    ): Response<AlertsResponse>
    
    @GET("/alerts/history")
    suspend fun getAlertHistory(): Response<AlertHistoryResponse>
    
    @POST("/alerts/{alertId}/dismiss")
    suspend fun dismissAlert(
        @Path("alertId") alertId: Int
    ): Response<Map<String, String>>
    
    // ============================================================
    // Health endpoints (pollen, flu, migraine, activities)
    // ============================================================
    
    @GET("/health/pollen")
    suspend fun getPollenForecast(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("days") days: Int = 5
    ): Response<PollenForecastResponse>
    
    @GET("/health/pollen/today")
    suspend fun getTodaysPollen(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<PollenTodayResponse>
    
    @GET("/health/flu-risk")
    suspend fun getFluRisk(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<FluRiskResponse>
    
    @GET("/health/migraine-risk")
    suspend fun getMigraineRisk(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<MigraineRiskResponse>
    
    @GET("/health/activities")
    suspend fun getAllActivities(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<AllActivitiesResponse>
    
    @GET("/health/activities/{activity}")
    suspend fun getActivityForecast(
        @Path("activity") activity: String,
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<ActivityDetailResponse>
    
    // ============================================================
    // Location endpoints
    // ============================================================
    
    @GET("/api/locations/search")
    suspend fun searchLocations(
        @Query("query") query: String,
        @Query("limit") limit: Int = 10
    ): Response<LocationSearchResponse>
    
    @GET("/api/locations/favorites")
    suspend fun getFavoriteLocations(): Response<FavoriteLocationsResponse>
    
    @POST("/api/locations/favorites")
    suspend fun addFavoriteLocation(
        @Query("name") name: String,
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("is_default") isDefault: Boolean = false
    ): Response<AddFavoriteResponse>
    
    @DELETE("/api/locations/favorites/{locationId}")
    suspend fun deleteFavoriteLocation(
        @Path("locationId") locationId: Int
    ): Response<Map<String, String>>
    
    @PATCH("/api/locations/favorites/{locationId}/default")
    suspend fun setDefaultLocation(
        @Path("locationId") locationId: Int
    ): Response<Map<String, String>>
    
    // ============================================================
    // Notification endpoints
    // ============================================================
    
    @POST("/notifications/register-device")
    suspend fun registerDevice(
        @Body body: Map<String, @JvmSuppressWildcards Any>
    ): Response<DeviceRegisterResponse>
    
    @POST("/notifications/unregister-device")
    suspend fun unregisterDevice(
        @Body body: Map<String, String>
    ): Response<Map<String, String>>
    
    @POST("/notifications/test")
    suspend fun sendTestNotification(
        @Body body: Map<String, String>
    ): Response<TestNotificationResponse>
    
    @GET("/notifications/preferences")
    suspend fun getNotificationPreferences(): Response<NotificationPreferencesResponse>
    
    @PUT("/notifications/preferences")
    suspend fun updateNotificationPreferences(
        @Body body: Map<String, Boolean>
    ): Response<NotificationPreferencesUpdateResponse>
    
    // ============================================================
    // Personalization endpoints
    // ============================================================
    
    @POST("/personalization/track")
    suspend fun trackEvent(
        @Body request: TrackEventRequest
    ): Response<Map<String, String>>
    
    // Every personalization route resolves the user from the bearer token. The
    // user_id query these used to send is not a parameter any of them declares,
    // so it was ignored — a caller passing one was not choosing a profile, it
    // just read its own.
    @GET("/personalization/profile")
    suspend fun getPersonalizationProfile(): Response<UserPreferenceProfile>

    @GET("/personalization/recommendations")
    suspend fun getPersonalizedRecommendations(
        @Query("temperature") temperature: Double,
        @Query("humidity") humidity: Int,
        @Query("uv_index") uvIndex: Double,
        @Query("precipitation_probability") precipProbability: Int
    ): Response<PersonalizedContentResponse>

    @GET("/personalization/should-notify")
    suspend fun shouldNotify(
        @Query("notification_type") notificationType: String
    ): Response<ShouldNotifyResponse>
    
    @POST("/personalization/profile/update")
    suspend fun updatePersonalizationProfile(
        @Body updates: Map<String, @JvmSuppressWildcards Any>
    ): Response<UserPreferenceProfile>
    
    // ============================================================
    // Precipitation Nowcast endpoint
    // ============================================================
    
    @GET("/api/v1/weather/nowcast")
    suspend fun getPrecipitationNowcast(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<PrecipitationNowcast>
    
    // ============================================================
    // Multi-Source Weather endpoints
    // ============================================================
    
    @GET("/api/weather/multi-source")
    suspend fun getMultiSourceWeather(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("sources") sources: String? = null
    ): Response<MultiSourceWeatherResponse>
    
    @GET("/api/weather/uv")
    suspend fun getUVIndex(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<UVIndexData>
    
    @GET("/api/weather/marine")
    suspend fun getMarineWeather(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double
    ): Response<MarineWeatherData>
    
    @GET("/api/weather/historical")
    suspend fun getHistoricalWeather(
        @Query("latitude") latitude: Double,
        @Query("longitude") longitude: Double,
        @Query("start_date") startDate: String,
        @Query("end_date") endDate: String
    ): Response<HistoricalWeatherData>
}

/**
 * Sign in with the two fields the caller actually holds.
 *
 * An extension rather than a member so that fixing the wire format above did not
 * ripple out into every call site: Retrofit needs the body as a single object,
 * but nobody calling this has one.
 */
suspend fun ClimaAIApi.login(email: String, password: String): Response<TokenResponse> =
    loginRequest(UserLogin(email = email, password = password))

// Body of PUT /api/auth/me. Every field on the router's UserUpdate is optional,
// so this only carries the part the app actually changes.
data class UserUpdateRequest(
    val preferences: UserPreferences
)

// Body of POST /api/subscriptions/activate — the router's SubscriptionCreate.
// platform and plan must be values of the backend's SubscriptionPlatform
// ("google"/"apple") and SubscriptionPlan ("monthly"/"annual") enums;
// receipt_data carries the Play purchase token.
data class SubscriptionActivateRequest(
    val platform: String,
    val plan: String,
    val receipt_data: String
)

// Request/Response classes for subscriptions
data class ReceiptValidationRequest(
    val platform: String,
    val receipt_data: String,
    val product_id: String? = null
)

data class ReceiptValidationResponse(
    val valid: Boolean,
    val platform: String,
    val is_active: Boolean?,
    val product_id: String?,
    val expires_at: String?,
    val error: String?
)

data class TrialRequest(val platform: String = "google")

data class PlansResponse(
    val plans: List<SubscriptionPlan>,
    val trial: TrialInfo
)

data class TrialInfo(
    val duration_days: Int,
    val features: String
)

object ApiClient {
    private var authToken: String? = null
    
    private val loggingInterceptor = HttpLoggingInterceptor().apply {
        level = if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor.Level.BODY
        } else {
            HttpLoggingInterceptor.Level.NONE
        }
    }
    
    private val okHttpClient = OkHttpClient.Builder()
        .addInterceptor(loggingInterceptor)
        .addInterceptor { chain ->
            val request = chain.request().newBuilder()
            authToken?.let {
                request.addHeader("Authorization", "Bearer $it")
            }
            chain.proceed(request.build())
        }
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val gson = GsonBuilder()
        .setDateFormat("yyyy-MM-dd'T'HH:mm:ss")
        .create()
    
    // In an unconfigured release build API_BASE_URL is the RFC 2606 sentinel
    // https://unconfigured.invalid/ (see build.gradle). That is safe to build
    // against: baseUrl() only parses the URL — no DNS, no connection — and
    // create() returns a lazy proxy, so this object never touches the network
    // until a call executes. Those calls fail fast with UnknownHostException,
    // which every repository already treats as offline. Keep it that way: no
    // eager health checks or warm-up pings against the base URL at init.
    private val retrofit = Retrofit.Builder()
        .baseUrl(BuildConfig.API_BASE_URL)
        .client(okHttpClient)
        .addConverterFactory(GsonConverterFactory.create(gson))
        .build()
    
    val api: ClimaAIApi = retrofit.create(ClimaAIApi::class.java)
    
    fun setAuthToken(token: String?) {
        authToken = token
    }
    
    fun getAuthToken(): String? = authToken
}
